content/blog/2025-12-18-extending-sql.md

```diff
@@ -0,0 +1,379 @@
+
+`PIVOT` rotates rows into columns, and `UNPIVOT` does the reverse. Neither requires new execution logic: `PIVOT` is just `GROUP BY` with `CASE` expressions, and `UNPIVOT` is a `UNION ALL` of each column. The planner rewrites them accordingly:
+
+```rust
+match relation {
+  TableFactor::Pivot { .. } => /* rewrite to GROUP BY + CASE */,
+  TableFactor::Unpivot { .. } => /* rewrite to UNION ALL */,
+  other => Original(other),
+}
+```
+
+Because the output is a standard `LogicalPlan`, DataFusion's usual optimization and physical planning apply automatically.
+
+**Full working example:** [`pivot_unpivot.rs`][pivot_unpivot.rs]
+
+### Strategy B: custom logical + physical (TABLESAMPLE)
+
+Sometimes rewriting is not sufficient. `TABLESAMPLE` returns a random subset of rows from a tableand is useful for approximations or debugging on large datasets. Because it requires runtime randomness, you cannot express it as a rewrite to existing operators. Instead, you need a custom logical node and physical operator to execute it.
+
+The approach (shown in [`table_sample.rs`][table_sample.rs]):
+
+1. `RelationPlanner` recognizes `TABLESAMPLE` and produces a custom logical node
+2. That node gets wrapped in `LogicalPlan::Extension`
+3. `ExtensionPlanner` converts it to a custom `ExecutionPlan`
+
+In code:
+
+```rust
+// Logical planning: FROM t TABLESAMPLE (...)  ->  LogicalPlan::Extension(...)
+let plan = LogicalPlan::Extension(Extension { node: Arc::new(TableSamplePlanNode { /* ... */ }) });
+```
+
+```rust
+// Physical planning: TableSamplePlanNode  ->  SampleExec
+if let Some(sample_node) = node.as_any().downcast_ref::<TableSamplePlanNode>() {
+  return Ok(Some(Arc::new(SampleExec::try_new(input, /* bounds, seed */)?)));
+}
+```
+
+This is the general pattern for custom FROM constructs that need runtime behavior.
+
+**Full working example:** [`table_sample.rs`][table_sample.rs]
+
+---
+
+## Putting it together
+
+When building your own dialect extension, work incrementally. First make it parse, then give it meaning, then (only if necessary) add custom execution.
+
+**Parsing** is the entry point. Wrap `DFParser`, intercept the keywords you care about, and produce either a standard `Statement` or your own `MyStatement`.
+
+**Meaning** comes from the planner traits. Which one depends on what you are adding: `ExprPlanner` for operators and functions, `TypePlanner` for data types, `RelationPlanner` for FROM-clause constructs.
+
+**Execution** is often unnecessary. Prefer rewrites to existing operators. If you do need custom runtime behavior (randomness, state, I/O), follow the `TABLESAMPLE` pattern and wrap a custom logical node in `LogicalPlan::Extension`, then convert it to an `ExecutionPlan` via `ExtensionPlanner`.
+
+---
+
+## Debugging tips
+
+### Print the logical plan
+
+```rust
+let df = ctx.sql("SELECT * FROM t TABLESAMPLE (10 PERCENT)").await?;
+println!("{}", df.logical_plan().display_indent());
+```
+
+### Use `EXPLAIN`
```

Comment:
Maybe worth adding a link to the docs: https://datafusion.apache.org/user-guide/sql/explain.html